Lib/pathlib/_local.py

```diff
@@ -780,6 +784,25 @@ def mkdir(self, mode=0o777, parents=False, exist_ok=False):
             if not exist_ok or not self.is_dir():
                 raise
 
+    if hasattr(_winapi, 'CopyFile2'):
+        def copy(self, target, follow_symlinks=True):
+            """
+            Copy the contents of this file to the given target. If this file is a
+            symlink and follow_symlinks is false, a symlink will be created at the
+            target.
+            """
+            try:
+                target = os.fspath(target)
+            except TypeError:
+                if isinstance(target, PathBase):
+                    # Target is an instance of PathBase but not os.PathLike.
+                    # Use generic implementation from PathBase.
+                    return PathBase.copy(self, target, follow_symlinks=follow_symlinks)
+                raise
+
+            flags = 0 if follow_symlinks else _winapi.COPY_FILE_COPY_SYMLINK
+            _winapi.CopyFile2(os.fspath(self), target, flags)
```

Comment:
`CopyFile2()` doesn't copy a directory symlink unless the flag `COPY_FILE_DIRECTORY` is passed. However, Python follows POSIX file types, for which a symlink cannot be a directory, and it's expected that any symlink can be copied. That said, we can't always use the `COPY_FILE_DIRECTORY` flag. It should be added only if `st_file_attributes` includes `stat.FILE_ATTRIBUTE_DIRECTORY` and `st_reparse_tag` is `stat.IO_REPARSE_TAG_SYMLINK`. This could also be expanded to include `stat.IO_REPARSE_TAG_MOUNT_POINT` to copy junctions, since `follow_symlinks` applies to junctions as well for the purposes of `os.lstat()`, and `os.unlink()` deletes junctions as if they're symlinks.

Reply:
Thanks Eryk, that's very helpful. I suppose there are three ways forward:

1. Perform an `os.path.isdir()` check first, and add `COPY_FILE_DIRECTORY` to the flags if it comes back true
2. Ignore the discrepancy. It looks like it already affects `shutil.copy2()` and the sky hasn't fallen in.
3. Remove the `follow_symlinks` argument and always follow symlinks.

I think I'm leaning towards #3 for simplicity's sake. We'd be able to add `follow_symlinks` later in response to user demand. What do you reckon?

Reply:
`isdir()` can't be used because returning `True` requires that the target path exists, so the `st_file_attributes` result from `lstat()` has to be checked. The system error code in this case is `ERROR_ACCESS_DENIED`, in which case `shutil.copy2()` uses the fallback implementation that calls `shutil.copyfile()`, which from the lack of commentary about this case seems like it just happened to work by a lucky accident that the error code is the same as for "a hidden or readonly file we can't overwrite". The implementation of `shutil.copyfile()` checks `follow_symlinks and _islink(src)`, in which case it executes `os.symlink(os.readlink(src), dst)`. We could handle it a bit like that, in which we check the `os.lstat()` result when handling `OSError`, except more narrowly focused. If it's a directory symlink, retry `CopyFile2()` with `COPY_FILE_SYMLINK | COPY_FILE_DIRECTORY`. This opens the door to copying junctions as well. That would bring the standard library a step closer to handling name-surrogate reparse points consistently for `follow_symlinks`.

Reply:
Thanks again. I've added an implementation along those lines, though my `is_dirlink()` function seems to be miscategorising symlinks-to-files as symlinks-to-directories. I'll look into it later.